core/src/main/java/hudson/logging/LogRecorder.java

```diff
@@ -77,13 +78,52 @@
 public class LogRecorder extends AbstractModelObject implements Saveable {
     private volatile String name;
 
-    public final CopyOnWriteList<Target> targets = new CopyOnWriteList<>();
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("TODO")
+    /**
+     * No longer used.
+     *
+     * @deprecated use {@link #getLoggers()}
+     */
+    public transient final CopyOnWriteList<Target> targets = new CopyOnWriteList<>();
+    private List<Target> loggers = new ArrayList<>();
     private final static TargetComparator TARGET_COMPARATOR = new TargetComparator();
-    
+
+    @DataBoundConstructor
+    public LogRecorder(String name) {
+        this.name = name;
+        // register it only once when constructed, and when this object dies
+        // WeakLogHandler will remove it
+        new WeakLogHandler(handler,Logger.getLogger(""));
+    }
+
+    private Object readResolve() {
+        if (loggers == null) {
+            loggers = new ArrayList<>();
+        }
+
+        if (!targets.isEmpty()) {
```

Comment:
the same is true for targets, this could also be null and would cause a null point exception.

```suggestion
        if (targets != null && !targets.isEmpty()) {
```